File: skills/docforge/_shared/runtime/common/python/flow_index_schema.py
"""Flow-index schema versioning shared by the flows runtime and metadata
migration.

Both `flow_index` and `migrate_metadata` need the same additive upgrade, so it
lives here in `common` rather than inside the flows runtime. Internal only: no
CLI launcher.
"""

from __future__ import annotations

FLOW_INDEX_VERSION = "1.2"
SUPPORTED_FLOW_INDEX_VERSIONS = ("1.1", "1.2")


def upgrade_index(index: dict) -> dict:
    """Additively upgrade a 1.1 flow index to the current schema.

    Mutates and returns the same dict. Raises ValueError on unsupported
    versions; a current index passes through untouched."""
    version = str(index.get("version") or "1.1")
    if version == FLOW_INDEX_VERSION:
        return index
    if version not in SUPPORTED_FLOW_INDEX_VERSIONS:
        raise ValueError(f"unsupported flow index version: {version}")
    index["version"] = FLOW_INDEX_VERSION
    if index.get("summary") and "written" not in index["summary"]:
        index["summary"]["written"] = sum(
            1
            for row in index.get("flows", [])
            if isinstance(row, dict)
            and row.get("status") == "documented"
            and bool(row.get("summary"))
        )
    return index

File: skills/docforge/_shared/runtime/common/js/flow_index_schema.js
"use strict";
/** Flow-index schema versioning shared by the flows runtime and metadata
 * migration.
 *
 * Both `flow_index` and `migrate_metadata` need the same additive upgrade, so
 * it lives here in `common` rather than inside the flows runtime. Internal
 * only: no CLI launcher.
 */

const FLOW_INDEX_VERSION = "1.2";
const SUPPORTED_FLOW_INDEX_VERSIONS = ["1.1", "1.2"];

function upgradeIndex(index) {
  // Additively upgrade a 1.1 flow index to the current schema. Mutates and
  // returns the same object; throws on unsupported versions.
  const version = String(index.version || "1.1");
  if (version === FLOW_INDEX_VERSION) return index;
  if (!SUPPORTED_FLOW_INDEX_VERSIONS.includes(version)) {
    throw new Error(`unsupported flow index version: ${version}`);
  }
  index.version = FLOW_INDEX_VERSION;
  if (index.summary && !("written" in index.summary)) {
    index.summary.written = (index.flows || [])
      .filter((row) => row && row.status === "documented" && Boolean(row.summary))
      .length;
  }
  return index;
}

module.exports = { FLOW_INDEX_VERSION, SUPPORTED_FLOW_INDEX_VERSIONS, upgradeIndex };
